File: 01-Arrays/main.js
// arrays in JavaScript are dynamic and can allow store different data types
// there is no fixed size array in java script
// array variables hold refrences adresses. Those refrence adress then point to the memory adress of the first elemetn in the list. Subsequent elemetents are stored in contgious blocks.

/*
Here is an example of a refrence:
array1 ---> Ref1234
array2  ---> Ref1234
Ref1234 ---> [1, 2, 3, 4, 5]
*/

let numbers = [1, 2, 3, 4, 5, 6];
// can access elements using their index 
console.log(numbers[0]);

// can modify indexes
numbers[0] = 10;

// arrays can store different types and objects. 
let mixedTypes = [1,2,"Dog", true];
console.log(mixedTypes);

let fruits = ['apple', 'orange', 'banana', 'coconut'];
console.log(fruits);

// this is how to add elements to the end of the list
fruits.push('coconut');

// how to remove elements from the end of the list
fruits.pop();

// how to add elements to the begining of the list 
fruits.unshift('mango');

// how to remove elements from the begining of the list
fruits.shift();

let numOfFruits = fruits.length // how to get length of an array

// how to search for the index of an element
let index = fruits.indexOf('apple');
console.log(index); // returns -1 if element is not found 

// use a for loop to iterate through the array
for (let i = 0; i < fruits.length; i++){
  console.log(fruits[i]); 
}

// for each loop
for (let fruit of fruits){
  console.log(fruit);
}


// another forEach loop (better)
fruits.forEach((fruit) => {
  console.log(fruit);
});

// how to sort the array
fruits.sort(); 
console.log(fruits);

// arrays are refrences as well. If you make a copy of an array you are making another variable but assigning it to the same memeory adress
// so chaning a copy of the array will affect the original array. 
let array2 = numbers;
array2.pop();
console.log(numbers);

// the splice method is used to add remove or replace elemetns in an array and returns the modified array
let arr1 = [1,2,3,4,5,6];
let removedArr1 = arr1.splice(2,2); // start at index 2 and delete 2 elements (the second element and one more). Store the deleted elments in modArr1.
console.log(arr1);
console.log(removedArr1);

arr1.splice(2,0,"a","b"); // start at index 2, remove 0 elements, add a and b
console.log(arr1);

let replacedArr1 = arr1.splice(2,2,3,4); // start at index 2, remove 2 elements (index 2 and the next element), and add numbers 3 and 4 starting from index 2. 
console.log(arr1);
console.log(replacedArr1); // holds the deleted items. 

// do not need to save the splice to a var can just call it on an array as well
arr1.splice(0,2);
console.log(arr1);

File: 01-Arrays/main.py
# Python has Lists (no arrays). Lists are dynamic (no fixed size), can hold multiple types, and are passed as refrences. They can be sorted. Duplicate values are ok. 

# create a list by surrounding data types with [] and seperating values with ","
fruits = ["apple", "orange", "banana", "coconut"]
print(fruits)
# this is how to access an element in a List
print(fruits[0])
# you can choose what index to start the list from
print(fruits[1:])
# can create a step that prints every second element
print(fruits[::2])
# can reverse the order of the list by using a -1 step
print(fruits[::-1])

# use len function to get the length of a list
print(len(fruits))

# can iterate over a list using a for loop
for x in range(len(fruits)):
  print(fruits[x])

# can use a for each loop as well
for fruit in fruits:
  print(fruit)

# if you forget what methods you can use on a list, print the dir() function
# print(dir(fruits))

# if you forget what each method does, you can print the help()
# print(help(fruits))

# you can check if a value is contained in a list
contain = "apple" in fruits
print(contain)

# can change the element at an index
fruits[0] = "grape"

# can add items to a list
fruits.append("pineapple")
print(fruits)

# can remove an item at a certain index or the item itself if it is in the list 
fruits.remove(fruits[1])
fruits.remove("pineapple")
print(fruits)

# can retrieve the index of a element
print(fruits.index("grape"))

# can insert an item at a specific index
fruits.insert(1, "apple")
print(fruits)

# can sort the list
fruits.sort()
print(fruits)

# can reverse a list 
fruits.reverse()
print(fruits)

# can save a sorted list to a variable
sortedFruits = sorted(fruits)
sortedFruits[3] = "bob"
print(sortedFruits)
print(fruits) # sorted fruits is a seperate list from fruits not pointing to the same refrence. Changes to sortedFruits does not affect fruits

# lists are refrences, changing the value of a copy will affect the original list
number = [1,2,3,4]
number2 = number
number2[0] = 10 # even if we modify number2 it will affect number1 since they point to the same refrence in memory.
print(number) 

#-----------------------------------------------------------------------------------------------------------------------------#

# tupes are ordered lists but are unchangeable. They can store multiple types. Duplicate values are allowed. Faster than a list. 

# declare a tuple by using () instead of []
friends = ('Bob', 'Sam', 'Tim')
print(friends)

# can check if a tuple contains an element
print("Tom" in friends)

# can keep track of how many of the same items are in a list using the count method
print(friends.count("Bob"))

# can iterate over a tuple
for friend in friends:
  print(friend)